tensorboard/webapp/reloader/reloader_component.ts

```diff
@@ -22,21 +22,34 @@ import {reload} from '../core/actions';
 
 /** @typehack */ import * as _typeHackRxjs from 'rxjs';
 
+const documentUtil = {
+  /**
+   * Wraps Page Visibility API call to determine if document is visible.
+   * Can be overriden for testing purposes.
+   */
+  getVisibilityState(): string {
+    return document.visibilityState;
+  },
+};
+
 @Component({
   selector: 'reloader',
   template: '',
   changeDetection: ChangeDetectionStrategy.OnPush,
 })
 export class ReloaderComponent {
+  private readonly onVisibilityChange = this.onVisibilityChangeImpl.bind(this);
   private readonly reloadEnabled$ = this.store.pipe(select(getReloadEnabled));
   private readonly reloadPeriodInMs$ = this.store.pipe(
     select(getReloadPeriodInMs)
   );
   private reloadTimerId: ReturnType<typeof setTimeout> | null = null;
+  private missedAutoReload: boolean = false;
 
   constructor(private store: Store<State>) {}
 
   ngOnInit() {
+    window.addEventListener('visibilitychange', this.onVisibilityChange);
```

Comment:
rxjs has utility called `fromEvent` and you can write this in more Rxjs way.

Pseudo code
```ts
fromEvent(window. 'visibilitychange').pipe(
  takeUntil(this.ngUnsubscribe),
  filter(() => {
    return documentUtil.getVisibilityState() === 'visible';
  }),
  tap(() => {
    this.store.dispatch(reload()),
  }),
);
```

Of course, since we do not really want hard dependency on DOM, you may also want to DI the `Document` helper or https://angular.io/api/common/DOCUMENT.

Reply:
I went with the approach of injecting the Document.  Since some of the test framework relies on the injected Document it isn't totally straightforward to fake in the tests and is a bit brittle.  Take a look at `fakeDocument` and let me know what you think.